plugins/woocommerce/src/Blocks/BlockTypes/ProductButton.php

```diff
@@ -213,17 +202,18 @@ protected function render( $attributes, $content, $block ) {
 
 		$div_directives = '
 			data-wp-interactive="woocommerce/product-button"
-			data-wp-context=\'' . wp_json_encode( $context, JSON_NUMERIC_CHECK | JSON_HEX_TAG | JSON_HEX_APOS | JSON_HEX_QUOT | JSON_HEX_AMP ) . '\'
 			data-wp-init="actions.refreshCartItems"
 		';
 
+		$context_directives = wp_interactivity_data_wp_context( $context );
+
 		$button_directives = $is_descendant_of_add_to_cart_form ?
 			'data-wp-class--disabled="woocommerce/add-to-cart-with-options::!state.isFormValid"' :
 			'data-wp-on--click="actions.addCartItem"';
 		$anchor_directive  = $is_descendant_of_add_to_cart_form ? '' : 'data-wp-on--click="woocommerce/product-collection::actions.viewProduct"';
 
 		$span_button_directives = '
-			data-wp-text="state.addToCartText"
+			data-wp-text="context.addToCartText"
```

Comment:
This should be 
```suggestion
			data-wp-text="state.addToCartText"
```


[According to the documentation](https://developer.wordpress.org/block-editor/reference-guides/interactivity-api/core-concepts/server-side-rendering/#:~:text=Let%E2%80%99s%20fix%20that,logic%20in%20PHP%3A), we need to define the derived state too on the server side to avoid any flickering effect. In this case, this snippet is enough.

```php
		wp_interactivity_state(
			'woocommerce/product-button',
			array(
				'addToCartText' => function () {
					$context = wp_interactivity_get_context();
					return $context['addToCartText'];
				},
			)
		);
```

Reply:
To prevent flickering (and to make it work without JS) we could also just provide text in the button element. That's what I was going for by removing the `$is_ajax_button` check. Adding state seems like overkill when we can do `<button>text</button>`. It's just a question of the right text, since as you pointed out, it's NOT "add to cart" in the product collection if the ajax setting is not enabled. :thinking:

Reply:
ah i see. if we go back to `state.addToCartText` (And that's needed to fix the button) then WP will parse that and we do have to provide an initial state. on it.

Reply:
> ah i see. if we go back to state.addToCartText (And that's needed to fix the button) then WP will parse that and we do have to provide an initial state. on it.

Yep, exactly!

Reply:
If the add to cart text is parsed by the iAPI server side state, then does `$is_ajax_button` still matter for the `{add_to_cart_text}`?

Reply:
No, we can remove `{add_to_cart_text}`.

Reply:
No. For this reason, I think that we need to include the `$is_ajax_button` check when the `add_to_cart_text` is defined in the context: does it make sense?

Reply:
Not yet :sweat_smile: if it's parsed by the iAPI in all cases, you can put text there or not and it's still going to be the state value. so what does the conditional do differently?